Migration of CameraAVSM cluster to code driven model

pidarped opened the pull request:
#### Summary
Migrate CameraAVSM cluster to code driven model

#### Related issues

#### Testing

#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [ ] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html#title-formatting)
-   [ ] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [ ] PR size is short
-   [ ] Try to avoid "squashing" and "force-update" in commit history
-   [ ] CI time didn't increase

See: [Pull Request Guidelines](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html)

pidarped committed:
Initial changes to moving CameraAVSM to code-driven model.

pidarped committed:
Modify command processing logic to return Status directly from
the command handling methods.

pidarped committed:
Modify camera-app initialization of CameraAVSM to use code-driven APIs.

pidarped committed:
Renames and fixes

pidarped committed:
Add CodegenIntegration.cpp

andy31415 commented on src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-cluster.cpp:
Please use `CameraAvStreamManagementCluster.cpp` as the file name, so that we have the logic of `class Foo` resides in `Foo.h`. 

We did this for existing code driven clusters already (and that was painful ... hoping the pain was only once)

pidarped committed:
Fix build errors and issues after rebase.

pidarped committed:
Restyle fixes

pidarped committed:
Add unit test for CameraAVSM.

pidarped committed:
Modify command handlers' arguments to remove unnecessary CommandHandler
and CommandPath arguments for some of the commands.

pidarped committed:
Replace MatterReportingAttributeChangeCallback with
NotifyAttributeChanged.

pidarped committed:
Restyle fixes

pidarped committed:
Add Register function in camera-app for CameraAVStreamMgmt.

pidarped committed:
Handle CHIP_ERROR from the setter functions during CameraAVSM init.

pidarped committed:
Remove CameraAVSM delegate stub from all-clusters-app.

The chip-camera-app is the de-facto implementation of the camera
application with the CameraAVSM delegate.

pidarped committed:
Add test functions for commands

pidarped committed:
Have the CameraAVSM cluster and ClusterTester be a member in the test
struct to be usable by each test case.

pidarped committed:
Add some more attribute units tests.

andy31415 commented on examples/camera-app/camera-common/src/camera-app.cpp:
```diff
@@ -51,6 +51,142 @@ CameraApp::CameraApp(chip::EndpointId aClustersEndpoint, CameraDeviceInterface *
+
+    ReturnErrorOnFailure(mAVStreamMgmtServer.Cluster().Init());
+
+    return CHIP_NO_ERROR;
```
```suggestion
    return mAVStreamMgmtServer.Cluster().Init();
```

yufengwangca commented on src/app/clusters/camera-av-stream-management-server/BUILD.gn:
```diff
@@ -11,5 +11,33 @@
+  sources = [
+    "CameraAVStreamManagementCluster.cpp",
+    "CameraAVStreamManagementCluster.h",
+    "CodegenIntegration.cpp",
```
should CodegenIntegration.cpp be defined in app_config_dependent_sources.gni?

pidarped replied:
Yes, this can be moved to the .gni file.

yufengwangca commented on src/app/clusters/camera-av-stream-management-server/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,29 @@
+    ChipLogProgress(Zcl, "Initializing Camera AVStream Management cluster.");
+}
+
+void MatterCameraAvStreamManagementPluginServerShutdownCallback()
```
It looks this file doe nothing except logging, if we initiate the cluster instance within app, do we still need CodegenIntegration.cpp?

pidarped committed:
Merged multiple AppendElements into single call.

pidarped committed:
Remove feature checks from InvokeCommand since AcceptedCommands already
performs those checks.

pidarped committed:
Incorporate review comments.

pidarped commented on src/app/clusters/camera-av-stream-management-server/CameraAVStreamManagementCluster.cpp:
```diff
@@ -1796,148 +1813,238 @@ CHIP_ERROR CameraAVStreamManagementCluster::LoadAllocatedStreams()
-            HandleCommand<Commands::VideoStreamDeallocate::DecodableType>(
-                handlerContext,
-                [this](HandlerContext & ctx, const auto & commandData) { HandleVideoStreamDeallocate(ctx, commandData); });
+            return Status::UnsupportedCommand;
```
Removed the checks for the commands in InvokeCommand.

pidarped commented on examples/camera-app/camera-common/src/camera-app.cpp:
```diff
@@ -51,6 +51,142 @@ CameraApp::CameraApp(chip::EndpointId aClustersEndpoint, CameraDeviceInterface *
+
+CHIP_ERROR CameraApp::InitializeCameraAVStreamMgmt()
+{
+    // Set the attribute defaults
```
Added.

pidarped commented on examples/camera-app/camera-common/src/camera-app.cpp:
```diff
@@ -142,138 +278,25 @@ CameraApp::CameraApp(chip::EndpointId aClustersEndpoint, CameraDeviceInterface *
+                               maxConcurrentVideoEncoders, maxEncodedPixelRate, sensorParams, nightVisionUsesInfrared, minViewport,
+                               rateDistortionTradeOffPoints, maxContentBufferSize, micCapabilities, spkrCapabilities,
+                               twowayTalkSupport, snapshotCapabilities, maxNetworkBandwidth, supportedStreamUsages,
+                               streamUsagePriorities);
```
Filed https://github.com/project-chip/connectedhomeip/issues/42573 to tackle this issue.

pidarped committed:
Add more attribute read tests to read initialized values.

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/CameraAVStreamManagementCluster.h:
```diff
@@ -686,8 +710,8 @@ class CameraAVStreamManagementCluster : public CommandHandlerInterface, public A
 
     CameraAVStreamManagementDelegate & mDelegate;
+    const BitFlags<Feature> mEnabledFeatures;
     EndpointId mEndpointId;
```
Why is this mEndpointId here?  We no longer set it in the constructor, so it just has a random value?

bzbarsky-apple commented on src/app/clusters/camera-av-stream-management-server/tests/TestCameraAVStreamManagementCluster.cpp:
I did not review this file, but see my comments above about things it should have caught and did not.

pidarped committed:
Remove Camera AVSM cluster from camera-app .zap and .matter files.

Also, add the cluster name to the CodeDriven clusters array in
config-data.yaml

pidarped committed:
Address review comments and remove unnecessary stub CodegenIntegration.cpp.

pidarped committed:
Add camera defaults init function and modify tests for corresponding
attributes to validate against those values.

pidarped committed:
Add the WaterMark feature to the test and add the Video and Snapshot
Modify commands.

pidarped commented on src/app/clusters/camera-av-stream-management-server/CameraAVStreamManagementCluster.cpp:
```diff
@@ -1135,15 +1142,14 @@ CHIP_ERROR CameraAVStreamManagementCluster::SetViewport(const Globals::Structs::
     mViewport = aViewport;
 
-    TEMPORARY_RETURN_IGNORED StoreViewport(mViewport);
+    ReturnErrorOnFailure(StoreViewport(mViewport));
```
Ok. Restored back for now.

pidarped commented on src/app/clusters/camera-av-stream-management-server/CameraAVStreamManagementCluster.cpp:
```diff
@@ -1796,148 +1813,197 @@ CHIP_ERROR CameraAVStreamManagementCluster::LoadAllocatedStreams()
+    if (HasFeature(Feature::kVideo) && (HasFeature(Feature::kWatermark) || HasFeature(Feature::kOnScreenDisplay)))
+    {
+        ReturnErrorOnFailure(builder.AppendElements(
+            { Commands::VideoStreamModify::kMetadataEntry, Commands::SnapshotStreamModify::kMetadataEntry }));
```
Yes, this was a mistake. Corrected. I think the unit test did not catch this because WMARK or OSD were not enabled as features. Enabled WMARK and added the Modify functions. 
Note the unit test would need more updates, e.g., for the commands which I intend to bring in a follow-up PR.

pidarped committed:
Add a README file to explain how the application should use this cluster
in its code-driven model and not use the ZAP tool for configuration.

andy31415 reviewed (approved):
Approving this after about 1 month of PR uptime and based on approval from @yufengwangca . Current assumption is that camera tests should cover at least part of this as well.
